fix `torch.nn.functional.l1_loss` for complex inputs

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65681
* #68419
* #69212
* #67079
* #67023
* #67022
* #69211

Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc @ezyang @gchanan @albanD @mruberry @jbschlosser @walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

pmeier committed:
fix `torch.nn.functional.l1_loss`

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -463,35 +455,10 @@ Tensor& mse_loss_backward_out(const Tensor& grad_output,
-  auto norm = reduction == Reduction::Mean ? grad_output / input.numel() : grad_output;
-  return at::sub_out(grad_input, input, target).sgn_().mul_(norm);
+  auto diff = input - target;
+  auto output = diff.is_complex() ? at::real(diff).abs() + at::imag(diff).abs()
+                                  : diff.abs();
```
If `v` is a complex or a real vector, the L1 norm is then `||v||_1 = \sum_{i=1}^n |v_i|`, where `|v_i|` is the module of its i-th coordinate. In case of complex numbers it is `sqrt(real^2 + imag^2)`.

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

ezyang commented:
I don't think removing the out variant is warranted. In the limit, every operator in pytorch should support out=. We may not necessarily prioritize it in all cases, but it's just a generally useful piece of surface area.

pmeier commented:
@ezyang 

> I don't think removing the out variant is warranted. In the limit, every operator in pytorch should support out=. We may not necessarily prioritize it in all cases, but it's just a generally useful piece of surface area.

So you suggest to put the out variant back and expose it in Python?

anjali411 commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -339,9 +339,6 @@ Tensor smooth_l1_loss(const Tensor& input, const Tensor& target, const int64_t r
 
 Tensor& smooth_l1_loss_out(const Tensor& input, const Tensor& target, int64_t reduction, double beta, Tensor& result) {
   TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
-  if (beta == 0) {
```
why do we wanna remove this?

pmeier replied:
We might be able to put it back after https://github.com/pytorch/pytorch/pull/65681#issuecomment-928836372 is resolved. For the other two cases, we can't keep them, because they call into functions that will no longer be available (`l1_loss_backward_out` and `l1_loss_backward`). 

Note that this is only a shortcut and does not change the output in case it is left out. It might be a perf regression though. I'm going to benchmark.

pmeier replied:
I've benchmarked with [this script]() and these are the results:

```
[---------- smooth_l1_loss with beta=0 ---------]
                             |  master  |    PR  
1 threads: --------------------------------------
      cpu / ()               |      30  |      24
      cpu / (10000000,)      |   56700  |   59000
      cpu / (10000, 10000)   |  626000  |  610000
      cuda / ()              |      30  |      25
      cuda / (10000000,)     |   60000  |   58000
      cuda / (10000, 10000)  |  612000  |  590000
2 threads: --------------------------------------
      cpu / ()               |      29  |      24
      cpu / (10000000,)      |   43000  |   49000
      cpu / (10000, 10000)   |  530000  |  490000
4 threads: --------------------------------------
      cpu / ()               |      29  |      24
      cpu / (10000000,)      |   39000  |   42000
      cpu / (10000, 10000)   |  500000  |  450000

Times are in microseconds (us).
```

- CUDA execution is faster than before
- CPU execution is slower for large 1D inputs, but faster for multi-dimensional inputs.

In all cases we are talking about a few milliseconds for quite large inputs.

ezyang commented:
> So you suggest to put the out variant back and expose it in Python?

Yes, though I'm not sure I'd necessarily do it in this PR. Depends on why it was easier to remove out in the first place here

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier commented:
@ezyang 

> > So you suggest to put the out variant back and expose it in Python?
> 
> Yes, though I'm not sure I'd necessarily do it in this PR. Depends on why it was easier to remove out in the first place here

I put the C++ out variant back, but did not expose it in Python yet. AFAIK, this would be the first `torch.nn.functional` operator with `out=` support so there should be a more general discussion how to handle this, for example do we only expose `out=` in the functional or also in the corresponding module? Thus, I would follow your advice and not do this in this PR.

ezyang commented:
> I put the C++ out variant back, but did not expose it in Python yet. AFAIK, this would be the first torch.nn.functional operator with out= support so there should be a more general discussion how to handle this, for example do we only expose out= in the functional or also in the corresponding module? Thus, I would follow your advice and not do this in this PR.

Yep, sgtm. Long term probably the right call is everything in nn.functional supports out.

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

anjali411 commented:
test failures look real:

```
Oct 05 08:26:37 ======================================================================
Oct 05 08:26:37 ERROR [0.010s]: test_torch_nn_functional_L1Loss_no_reduce_complex (__main__.TestCppApiParity)
Oct 05 08:26:37 ----------------------------------------------------------------------
Oct 05 08:26:37 Traceback (most recent call last):
Oct 05 08:26:37   File "/var/lib/jenkins/workspace/test/cpp_api_parity/functional_impl_check.py", line 202, in test_fn
Oct 05 08:26:37     unit_test_class=self, test_params=unit_test_class.functional_test_params_map[self._testMethodName])
Oct 05 08:26:37   File "/var/lib/jenkins/workspace/test/cpp_api_parity/functional_impl_check.py", line 87, in test_forward
Oct 05 08:26:37     python_output = run_forward(unit_test_class, test_params)
Oct 05 08:26:37   File "/var/lib/jenkins/workspace/test/cpp_api_parity/functional_impl_check.py", line 75, in run_forward
Oct 05 08:26:37     python_output = test_params.test_instance.constructor()(*inputs)
Oct 05 08:26:37   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 1106, in _call_impl
Oct 05 08:26:37     return forward_call(*input, **kwargs)
Oct 05 08:26:37   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_nn.py", line 366, in forward
Oct 05 08:26:37     return fn(*args, **kwargs)
Oct 05 08:26:37   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_nn.py", line 605, in <lambda>
Oct 05 08:26:37     lambda i: F.l1_loss(i, t.type_as(i), reduction='none')),
Oct 05 08:26:37   File "/opt/conda/lib/python3.6/site-packages/torch/nn/functional.py", line 3128, in l1_loss
Oct 05 08:26:37     return torch._C._nn.l1_loss(input, target, _Reduction.get_enum(reduction))
Oct 05 08:26:37 RuntimeError: The size of tensor a (24) must match the size of tensor b (4) at non-singleton dimension 2
Oct 05 08:26:37 
Oct 05 08:26:37 ----------------------------------------------------------------------
Oct 05 08:26:37 Ran 1020 tests in 9.402s
Oct 05 08:26:37 
Oct 05 08:26:37 FAILED (errors=1, skipped=510, expected failures=2)
Oct 05 08:26:37 
Oct 05 08:26:37 Generating XML reports...
Oct 05 08:26:37 Generated XML report: test-reports/python-unittest/test_cpp_api_parity/TEST-TestCppApiParity-20211005082627.xml
Oct 05 08:26:37 Traceback (most recent call last):
Oct 05 08:26:37   File "test/run_test.py", line 1030, in <module>
Oct 05 08:26:37     main()
Oct 05 08:26:37   File "test/run_test.py", line 1008, in main
Oct 05 08:26:37     raise RuntimeError(err_message)
Oct 05 08:26:37 RuntimeError: test_cpp_api_parity failed!
```

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

lezcano reviewed (commented):
Two small things.
Also, please add the out versions of the backward operators to the lists in `test/backward_compatibility/check_backward_compatibility.py`.

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss`"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier commented:
This PR got the same treatment as #65469. The `OpInfo` was split off into #69211.

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc ezyang gchanan albanD mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc ezyang gchanan albanD mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc ezyang gchanan albanD mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc ezyang gchanan albanD mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:
```diff
@@ -90,16 +91,50 @@ static void addcdiv_cpu_kernel(TensorIteratorBase& iter, const Scalar& value) {
+        [norm_val_vec, zero_vec](Vec input, Vec target, Vec grad_output) -> Vec {
+          const auto x = input - target;
+          const auto y = norm_val_vec * grad_output;
+          return Vec::blendv(y.neg(), y, x > zero_vec);
```
Forgot the zero case here. This should fix it.
```suggestion
          return Vec::blendv(Vec::blendv(y.neg(), y, x > zero_vec), zero_vec, x == zero_vec);
```

Lezcano commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1495,6 +1472,29 @@ Tensor smooth_l1_loss_double_backward(const Tensor & grad, const Tensor & input,
+    const auto z = input - target;
+    const auto ret = sgn_backward(at::sgn(z), grad, z) * grad_output;
+    if (reduction == at::Reduction::Mean) {
+      const auto n = c10::multiply_integers(at::infer_size(input.sizes(), target.sizes()));
+      return ret / n;
```
```suggestion
      return ret / z.numel();
```

lezcano reviewed (commented):
Just found a couple small things when re-reading through this.
I also wonder whether are we testing the broadcasting case. I am not sure the double backward would work in that case...

ezyang commented:
bumping

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc ezyang gchanan albanD mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier committed:
Update on "fix `torch.nn.functional.l1_loss` for complex inputs"


Fixes #65683. In addition this removes the `out` variant for two reasons:

1. It is not exposed in Python: https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L3101-L3107
2. By using `out=`, the function becomes non-differential by design, which is not really helpful for a loss function.

cc ezyang gchanan albanD mruberry jbschlosser walterddr

Differential Revision: [D31730000](https://our.internmc.facebook.com/intern/diff/D31730000)

[ghstack-poisoned]

pmeier commented:
I intentionally left in some merge conflicts originating from #62558. @nikitaved wanted to handle them.

Lezcano commented:
After a discussion with @jbschlosser, this PR is superseded by https://github.com/pytorch/pytorch/pull/78257, so it can be closed.
